File: MyCommunities/views/recommend.js
const mysql = require('mysql');
const axios = require('axios');
const cosineSimilarity = require('compute-cosine-similarity');

const connection = mysql.createConnection({
  host: 'localhost',
  user: 'root',
  password: 'Mysql@123',
  database: 'my_communities'
});

connection.connect();

const getEmbedding = async (text) => {
  try {
    const response = await axios.post('http://localhost:8000/get_embedding', { text: text });
    return response.data.vector;
  } catch (error) {
    console.error('Error fetching embedding:', error);
    return null;
  }
};

const fetchData = () => {
  return new Promise((resolve, reject) => {
    connection.query('SELECT * FROM community', (error, communities) => {
      if (error) {
        return reject(error);
      }

      connection.query('SELECT * FROM communityusers', (error, userInteractions) => {
        if (error) {
          return reject(error);
        }

        resolve({ communities, userInteractions });
      });
    });
  });
};

const prepareEmbeddings = async (communities) => {
  const embeddings = await Promise.all(
    communities.map(async (community) => {
      const embedding = await getEmbedding(community.Description);
      community.Embedding = embedding;
      return embedding;
    })
  );
  return embeddings;
};

const recommendCommunities = async (searchDescription, communities, embeddings, topN = 3) => {
  const searchEmbedding = await getEmbedding(searchDescription);

  const similarities = embeddings.map((embedding) => cosineSimilarity(searchEmbedding, embedding));

  const similarIndices = similarities
    .map((similarity, index) => ({ similarity, index }))
    .sort((a, b) => b.similarity - a.similarity)
    .slice(0, topN)
    .map((item) => item.index);

  return similarIndices.map((index) => communities[index].Community_Name);
};

const main = async () => {
  try {
    const { communities, userInteractions } = await fetchData();

    const embeddings = await prepareEmbeddings(communities);

    const searchDesc = 'office wear';
    const recommendedCommunities = await recommendCommunities(searchDesc, communities, embeddings);

    console.log('Recommended Communities:');
    console.log(recommendedCommunities);
  } catch (error) {
    console.error('Error:', error);
  } finally {
    connection.end();
  }
};

main();

File: MyCommunities/public/recommend.py
from flask import Flask, request, jsonify
import mysql.connector
import pandas as pd
import spacy
from sklearn.metrics.pairwise import cosine_similarity

app = Flask(__name__)

nlp = spacy.load('en_core_web_md')

def get_embedding(text):
    doc = nlp(text)
    return doc.vector

def fetch_data():
    try:
        conn = mysql.connector.connect(
            host='localhost',
            user='root',
            password='Mysql@123',
            database='my_communities'
        )
        query_communities = "SELECT * FROM community"
        communities = pd.read_sql(query_communities, conn)
        query_user_interactions = "SELECT * FROM communityusers"
        user_interactions = pd.read_sql(query_user_interactions, conn)
    except mysql.connector.Error as err:
        print(f"Error: {err}")
        return None, None
    finally:
        if conn.is_connected():
            conn.close()
    return communities, user_interactions

def prepare_embeddings(communities):
    communities['Embedding'] = communities['Description'].apply(lambda x: get_embedding(x))
    embeddings = communities['Embedding'].tolist()
    return embeddings

def recommend_communities(search_description, communities, embeddings, top_n=3):
    search_embedding = get_embedding(search_description)
    similarities = [cosine_similarity([search_embedding], [emb])[0][0] for emb in embeddings]
    similar_indices = sorted(range(len(similarities)), key=lambda i: similarities[i], reverse=True)[:top_n]
    recommendations = communities.iloc[similar_indices]
    return recommendations[['Community_Name']]

@app.route('/recommend', methods=['GET'])
def recommend():
    query = request.args.get('query', '')
    communities, _ = fetch_data()
    embeddings = prepare_embeddings(communities)
    recommended_communities = recommend_communities(query, communities, embeddings)
    return jsonify(recommended_communities.to_dict(orient='records'))

@app.route('/favicon.ico')
def favicon():
    return '', 204

if __name__ == '__main__':
    app.run(port=5000, debug=True)
